include/seqan3/io/stream/iterator.hpp

```diff
@@ -166,6 +172,213 @@ class fast_istreambuf_iterator
+     * \param[in] rng The range to write.
+     * \returns If `range_type` models `std::ranges::forwarding_range` returns an iterator pointing to end of the range
+     *          (rng) else returns `void`.
+     *
+     * This function avoids the buffer-at-end check by writing the range in chunks, where a chunks has the size of
+     * the remaining space in the put area of the buffer.
+     * If the range type models `std::ranges::sized_range` the chunks are written using `std::ranges::copy_n`, which
+     * may use memcpy if applicable. Otherwise, a simple for loop iterates over the chunk.
+     *
+     * \attention You can only use the return value (end iterator) if your range type models
+     *            `std::ranges::forwarding_range`.
+     *
+     * Example:
+     *
+     * \include test/snippet/io/detail/iterator_write_range.cpp
+     */
+    template <std::ranges::forward_range range_type>
+    //!\cond
+        requires forwarding_range<range_type>
+    //!\endcond
+    auto write_range(range_type && rng)
+    {
+        using sen_t = std::ranges::sentinel_t<range_type>;
+        using it_t = std::ranges::iterator_t<range_type>;
+
+        it_t rng_it = std::ranges::begin(rng);
+        sen_t rng_end = std::ranges::end(rng);
+
+        size_t const buffer_space = stream_buf->epptr() - stream_buf->pptr();
+
+        if constexpr (std::ranges::sized_range<range_type>)
+        {
+            size_t const characters_to_write = std::min<size_t>(std::ranges::size(rng), buffer_space);
+            auto copy_res = std::ranges::copy_n(rng_it, characters_to_write, stream_buf->pptr());
+            rng_it = copy_res.in;
+            stream_buf->pbump(characters_to_write);
+        }
+        else
+        {
+            for (size_t i = 0; rng_it != rng_end && i < buffer_space; ++rng_it, ++i)
+            {
+                *stream_buf->pptr() = *rng_it;
+                stream_buf->pbump(1);
+            }
+        }
+
+        if (rng_it == rng_end) // no more characters to write
+            return rng_it;
+
+        // Push one more character and flush
+        if (stream_buf->overflow(*rng_it) == traits_t::eof())
+            throw std::ios_base::failure{"Cannot write to output stream (reached traits::eof() condition)."};
+
+        ++rng_it; // drop 1 character that has been written in overflow()
+
+        // Write remaining range recursively. For sized ranges, ensure that the subrange is also sized.
+        if constexpr (std::ranges::sized_range<range_type>)
+        {
+            // Note: Using ranges here because of bug https://github.com/ericniebler/range-v3/pull/1415 (fixed)
+            // TODO: change to std::ranges::subrange, when updating to range-v3 > 0.10.0
+            using subrange_t = ranges::subrange<it_t, sen_t, ranges::subrange_kind::sized>;
+            size_t const remaining_size = std::ranges::size(rng) - buffer_space - 1;
+            return write_range(subrange_t{rng_it, rng_end, remaining_size});
```

Comment:
I just saw this, but why do you recurse (i.e. build up a call-stack) instead of a loop?

Reply:
Yes you are right, I changed it into a for loop

Reply:
This also makes the subrange workaround obsolete